smart-contracts/contracts/cl-vault/src/vault/deposit.rs

```diff
@@ -36,90 +41,218 @@ pub(crate) fn _execute_any_deposit(
     unimplemented!()
 }
 
-/// Try to deposit as much user funds as we can into the a position and
-/// refund the rest to the caller
+/// allocate as many user funds as we can to the current positions, after adding to those positions
 pub(crate) fn execute_exact_deposit(
-    deps: DepsMut,
+    mut deps: DepsMut,
     env: Env,
     info: MessageInfo,
     recipient: Option<String>,
 ) -> Result<Response, ContractError> {
     // Unwrap recipient or use caller's address
     let recipient = recipient.map_or(Ok(info.sender.clone()), |x| deps.api.addr_validate(&x))?;
+    CURRENT_DEPOSITOR.save(deps.storage, &recipient)?;
 
-    let position_id = (POSITION.load(deps.storage)?).position_id;
-    let position = ConcentratedliquidityQuerier::new(&deps.querier)
-        .position_by_id(position_id)?
-        .position
-        .ok_or(ContractError::PositionNotFound)?
-        .position
-        .ok_or(ContractError::PositionNotFound)?;
+    debug!(deps, "depositing 1");
+    let positions = get_positions(deps.storage, &deps.querier)?;
+
+    debug!(deps, "depositing 2");
 
     let pool = POOL_CONFIG.load(deps.storage)?;
     let (token0, token1) = must_pay_one_or_two(&info, (pool.token0, pool.token1))?;
 
-    CURRENT_DEPOSIT.save(
-        deps.storage,
-        &CurrentDeposit {
-            token0_in: token0.amount,
-            token1_in: token1.amount,
-            sender: recipient,
-        },
+    let spot_price = get_spot_price(deps.storage, &deps.querier)?;
+
+    let psf = allocate_funds_per_position(
+        deps.branch(),
+        positions.clone(),
+        spot_price,
+        token0.amount,
+        token1.amount,
     )?;
 
-    // Create coins_to_send with no zero amounts
-    let mut coins_to_send = vec![];
-    if !token0.amount.is_zero() {
-        coins_to_send.push(token0.clone());
-    }
-    if !token1.amount.is_zero() {
-        coins_to_send.push(token1.clone());
-    }
+    let position_funds = psf.iter().zip(positions);
+
+    // sum up the amount that allocate_funds_per_position allocated to each position
+    let mut total_allocated = (Uint128::zero(), Uint128::zero());
+
+    // create a position for each position
+    let msgs: Result<Result<Vec<SubMsg>, ContractError>, ContractError> = position_funds
+        .map(|((pos1, amount0, amount1), (pos2, fp))| {
+            if pos1.position_id != pos2.position_id {
+                // this would be a programming error
+                panic!("Position ids in the zipped iterator don't match")
+            }
+
+            // we want to make sure deposit 1 is item 1 in the queue, deposit 2 item 2 in the queue
+            // TODO can this be less complex?
+            CURRENT_DEPOSIT.push_back(
```

Comment:
Current use of CURRENT_DEPOSIT look works, but for clarity we can also use one more state variable for pushing back the updated deposit in the handle_deposit_create_position_reply, which is finally used in the mint operation.